core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -577,14 +582,42 @@ private SnapshotInfo inProgressSnapshot(SnapshotsInProgress.Entry entry) {
                         shardSnapshotStatus.updateStage(IndexShardSnapshotStatus.Stage.FAILURE);
                         shardSnapshotStatus.failure(shardFailure.reason());
                         shardStatus.put(shardId, shardSnapshotStatus);
+                        hasFailedShards = true;
                     } else {
-                        IndexShardSnapshotStatus shardSnapshotStatus =
-                            repository.getShardSnapshotStatus(snapshotInfo.snapshotId(), snapshotInfo.version(), indexId, shardId);
-                        shardStatus.put(shardId, shardSnapshotStatus);
```

Comment:
So, the problem here is that these calls are always failing when the snapshot is in a FAILED state, right? So, wouldn't it it make sense to not even try to call getShardSnapshotStatus if snapshot failed by just wrapping these 3 lines into a an if statement that checks if the snapshot has a failed status and simply generate the fake shard status you are generating in line 609 below.